### DISCUSSION TITLE
Document upgrade/downgrade processes

### DISCUSSION THREAD

**Discussion by zamazan4ik** [Unleash/unleash#2546] [Q&A] [answered]
After reading the documentation I have the following questions regarding upgrade/downgrade processes:

- Is it possible to upgrade without downtime?
- How the downgrade should be performed? Is the downgrade process supported?
- Is it possible to downgrade without downtime?

Also, could be quite useful to know about compatibility guarantees between releases.

**Comment by Tymek** [comment, marked as answer]
It is possible only if you spin up another instance while the old one is still running and move traffic. We don't have it out of the box, since it's just a NodeJS app. Migrations, both up and down, are automatic (we had one issue in the past, 4.12, this was patched after a day). All data related to new features can be lost when downgrading.

> compatibility guarantees between releases

We follow semantic versioning. Only in major releases you can expect breaking changes. We provide [migration guides](https://docs.getunleash.io/reference/deploy/migration-guide) when needed.

**Reply by Tymek** [reply]
CC @thomasheartman

**Reply by thomasheartman** [reply]
Thanks for the CC, @Tymek! I think you covered this quite well; I don't have anything else to add 🙌🏼

**Reply by Tymek** [reply]
actually, I would like to add something 😅 

You can have a setup using [Unleash Proxy](https://docs.getunleash.io/reference/unleash-proxy) where it's OK if your main instance and DB are down for maintenance. SDKs also cache the toggles. Depending of course on requirements. If you mostly care about end-user SLA it's a fairly resilient and scalable system.

**Comment by mbsimonovic** [comment]
Hey guys, would be great to document the regular (w/ downtime) upgrade process for the helm chart installation. Just tried it from 2.7.2 to 2.8.0 and it fails because postgresql version changed and requires db upgrade. 

I tried to perform the upgrade and keep the postgresql version 11 as documented https://docs.bitnami.com/kubernetes/infrastructure/postgresql/administration/upgrade/, but it also failed because app couldn't connect to the db although the db pod was running, not sure where it went wrong...

**Reply by thomasheartman** [reply]
Oh, interesting. @chriswk, do you know anything about this?

**Reply by thomasheartman** [reply]
Hey, @mbsimonovic 👋🏼 

Firstly, I apologize for the delayed response. We acknowledge that the update to 2.8.0 ideally should have been marked as a major update. That was an oversight on our part, and I appreciate your understanding.

Regarding your setup, did you manage to get everything running smoothly? If you're using the embedded postgres, I'd like to point out that we currently don't have a direct upgrade path. It also doesn't have any backups and isn't optimized for persistent storage. Therefore, we advise running a separate instance of postgres instead, to ensure a smoother experience.

Let me know if you need any further assistance or guidance!

**Reply by mbsimonovic** [reply]
Hey no worries, just wanted to point out a potential issue. I've moved from embedded to a managed postgres instance so no problems for me anymore.

**Reply by thomasheartman** [reply]
Great to hear! And yeah, we appreciate it. I'll have a look and make sure to call the issues with the embedded postgres out explicitly in the docs if it isn't already there.
